### ISSUE TITLE
"From Position" breaks after castling

### ISSUE THREAD

**Issue by imBigo** [ShailChoksi/lichess-bot#234]
The bot tries to play as the opposite side when it is playing "From Position" and usually happens to me if it is completely losing. Evaluation will go from -2000 to 2000 and it will get stuck trying to play from the winning side. I am using Lc0.
```
Traceback (most recent call last):
  File "D:\LichessBot\logging_pool.py", line 16, in __call__
    result = self.__callable(*args, **kwargs)
  File "C:\Users\user\AppData\Local\Programs\Python\Python39\lib\site-packages\backoff\_sync.py", line 99, in retry
    ret = target(*args, **kwargs)
  File "D:\LichessBot\lichess-bot.py", line 201, in play_game
    li.make_move(game.id, best_move)
  File "D:\LichessBot\lichess.py", line 74, in make_move
    return self.api_post(ENDPOINTS["move"].format(game_id, move))
  File "C:\Users\user\AppData\Local\Programs\Python\Python39\lib\site-packages\backoff\_sync.py", line 99, in retry
    ret = target(*args, **kwargs)
  File "D:\LichessBot\lichess.py", line 64, in api_post
    response.raise_for_status()
  File "C:\Users\user\AppData\Local\Programs\Python\Python39\lib\site-packages\requests\models.py", line 941, in raise_for_status
    raise HTTPError(http_error_msg, response=self)
requests.exceptions.HTTPError: 400 Client Error: Bad Request for url: https://lichess.org/api/bot/game/hcLy9S2c/move/b1a3
```

This is the game where this problem occurred: 
https://lichess.org/hcLy9S2c/black

**Comment by imBigo** [comment]
I have actually found out that the problem lies in castling. Whenever the bot or the other player castles, the bot tries to play a move as the opposite side and will crash.

**Comment by imBigo** [comment]
I have found a fix for this issue for those using Leela. Modify the file **lichess-bot.py** so that you trick Leela into thinking the game is a chess960 game. This will allow Leela to send the correct form of castling. (Changes are in bold)
`def setup_board(game):
    if game.variant_name.lower() == "chess960":
        board = chess.Board(game.initial_fen, chess960=True)
    elif game.variant_name == "From Position":
        board = chess.Board(game.initial_fen, **chess960=True**)
    else:
        VariantBoard = find_variant(game.variant_name)
        board = VariantBoard()
    moves = game.state["moves"].split()
    for move in moves:
        board = update_board(board, move)

    return board`

**Comment by cubesdev** [comment]
is there any other fix possible without the need for chess960 being set to true? (the engine I want to use does not support chess960 x-fen)

**Comment by imBigo** [comment]
> is there any other fix possible without the need for chess960 being set to true? (the engine I want to use does not support chess960 x-fen)

Does the same issue happen with your engine?

**Comment by cubesdev** [comment]
> Does the same issue happen with your engine?

Yes. It is not depending on the used engine. The lichess-bot UCI Interface simply stops to send any further moves after any player castles. So the bot gets an outdated board representation (sometimes even in the opponent perspective) which breaks everything.

**Comment by imBigo** [comment]
> > Does the same issue happen with your engine?
> 
> Yes. It is not depending on the used engine. The lichess-bot UCI Interface simply stops to send any further moves after any player castles. So the bot gets an outdated board representation (sometimes even in the opponent perspective) which breaks everything.

I'm not 100% sure about this but my theory is that lichess defaults castling rights the way they are handled in chess960 in UCI;
`Optionally supports chess960. In Chess960, castling moves are encoded by a king move to the corresponding rook square. Use chess.Board.from_chess960_pos() to create a board with one of the Chess960 starting positions.` [https://python-chess.readthedocs.io/en/latest/core.html#board](https://python-chess.readthedocs.io/en/latest/core.html#board)
This is probably because you could have castling be possible even though the king is not on the 'e' file.

**Comment by cubesdev** [comment]
It is definitely a bug in lichess-bot code. It is unable to get/parse the current game state (list of moves) from lichess when castling happend in from_position mode. 

But yes, maybe it has something to do with the chess960 castling move encoding. Probably lichess encodes it that way in from_position and lichess-bot assumes traditional notation... but that's just a guess.

**Comment by imBigo** [comment]
Your best bet is probably to add support for chess960 castling for your engine since it did fix the issue for me using Leela.

**Comment by cubesdev** [comment]
Yes, this should work along with your lichess-bot.py fix. Thank you for figuring it out and sharing it here.

**Comment by imBigo** [comment]
You're welcome! Good luck with the fix 🙂

**Comment by hi-ogawa** [comment]
I use this workaround https://github.com/hi-ogawa/lichess-bot/commit/5d99ca2a32ee5c082fd891f09591b890d0bb661a since my toy engine doesn't support Chess960 option.

**Comment by cubesdev** [comment]
> I use this workaround [hi-ogawa@5d99ca2](https://github.com/hi-ogawa/lichess-bot/commit/5d99ca2a32ee5c082fd891f09591b890d0bb661a) since my toy engine doesn't support Chess960 option.

This workaround works great! It should be merged to the official lichess-bot.

**Comment by hi-ogawa** [comment]
I'm glad that helped. But I don't think my patch is PR-worthy since lichess considers "from position" same as "chess960" as in https://github.com/ornicar/scalachess/blob/dc9964e087bd4fe0c230ff8b1f553498ff5533ec/src/main/scala/format/UciDump.scala#L26. I guess legit engine should support chess960 option and It would makes more sense to merge imBigo's solution https://github.com/ShailChoksi/lichess-bot/issues/234#issuecomment-766454910 into master.

**Comment by MarkZH** [comment]
Is everybody experiencing this bug using the most recent version lichess-bot? A lot of code has changed since this was first reported, including upgrading the python-chess library.

It would also be very helpful to post logs from running lichess-bot with the `-v` option from when this bug occurs.

**Comment by MarkZH** [comment]
I just tried to play a From Position game with castling and I did not encounter this bug. https://lichess.org/v1563PDYdOc5

**Comment by imBigo** [comment]
> Thanks to everyone here for investigating this!
> 
> Is everybody experiencing this bug using the most recent version lichess-bot? A lot of code has changed since this was first reported, including upgrading the python-chess library. This problem might have been solved already.
> 
> It would also be very helpful to post logs from running lichess-bot with the `-v` option from when this bug occurs.
> 
> I worry about this fix because it would break every engine that does not support chess960 play (including mine).

Did the bug ever happen to your engine before switching to newest version? I have switched to new version but I have been having some issue getting my bot to play moves. I'll get back to you when I get the time to also test it. 
Glad to see it's working normally. Hopefully we don't need to use my fix.

**Comment by MarkZH** [comment]
This issue was fixed in the python-chess library ([bug report](https://github.com/ornicar/lila/issues/7750)). Lichess sends a short castle by black as `e8h8` in From Position games. This was misinterpreted as an illegal move by the old python-chess library until the latest update. Just now [I played a game](https://lichess.org/LdnMpDHYi6qF) with From Position while recording both the messages from lichess to lichess-bot and from lichess-bot to my engine.
```
2021-03-14 23:47:31,755: {'type': 'gameState', 'moves': 'b5e2 g8f6 b1c3 e7e6 e1f1 f8b4 a2a3 e8h8', 'wtime': 177640, 'btime': 174110, 'winc': 0, 'binc': 0, 'wdraw': False, 'bdraw': False, 'status': 'started'}
... omitted ...
2021-03-14 23:47:31,757: <UciProtocol (pid=1076281)>: << position fen r1bqkbnr/1ppppppp/p1n5/1B6/4P3/8/PPPP1PPP/RNBQK1NR w KQkq - 0 1 moves b5e2 g8f6 b1c3 e7e6 e1f1 f8b4 a2a3 e8g8
```
Notice that lichess sends `e8h8` for the castling move, but this is translated into `e8g8` before being sent to the engine.

In other words, the bug was real, but it was fixed by upgrading python-chess. No change to lichess-bot is needed.

**Comment by MarkZH** [comment]
@imBigo Just to make sure, after you upgraded lichess-bot, did you rerun `python3 -m pip install -r requirements.txt` to get the latest version of python-chess?

**Comment by hi-ogawa** [comment]
I believe the bug was in these lines https://github.com/ShailChoksi/lichess-bot/pull/259/files#diff-e326008cac463869bad7488e0e9b058e652abe5c59e980c0476f05bb86b10508L422-L423.
But MarkZH has refactored out those, so I believe the issue is gone now in the current master.

**Comment by imBigo** [comment]
> @imBigo Just to make sure, after you upgraded lichess-bot, did you rerun `python3 -m pip install -r requirements.txt` to get the latest version of python-chess?

I did, I had the bot running for a bit but now it's no longer playing moves. I'll try to figure this out tomorrow. I'll also close my pull request.

**Comment by imBigo** [comment]
Issue was fixed with #260
